### PR TITLE
moving json-rpc-middleware-stream from merged-packages to actual pack…

### LINKED ISSUE #1920: Moving json-rpc-middleware-stream from merged-packages into packages
1. The big leap
 Move migration target from merged-packages/ to packages/.
 Run yarn install in the root directory.
 Check that all tests are passing in migration target by running yarn workspace @metamask/<package-name> test.
2. Update downstream repos
 Add tsconfig reference paths for migration target in downstream packages and root.
 Bump migration target version in downstream packages and root.
Notes on why this version bump needs to happen as part of this PR (import module shadowing): eth-json-rpc-provider migration - Integration into packages/ #1738 (comment)
3. Linter fixes
 Apply yarn constraints fixes to migration target package.json file: yarn constraints --fix (run twice).
 Identify validator fixes for CHANGELOG using yarn workspace @metamask/<package-name> changelog:validate and apply the diffs.
4. Resolve or TODO downstream errors
If introducing the migration target breaks any downstream repos:
 Resolve simple errors
 Mark and ignore complex/blocked errors using @ts-expect-error TODO: annotations.
 Create a separate issue for resolving the marked errors as soon as the migration is completed.
Fix provider, middleware type errors in NetworkController and downstream packages #1823
5. Finalize merge
 Check that all tests are passing in all subpackages of core and CI.
 Merge packages/<package-name> directory into core main branch.

### PR DESCRIPTION
**Pull request by kanthesha**
## Explanation

We chose to place json-rpc-middleware-stream initially in the merged-packages and not packages so we can hide it from Yarn, ESLint, Prettier, and TypeScript. In this PR, we cleaned it up and moved it from `merged-packages` to actual `packages` folder.

In this process, we followed below steps.
Moving to packages
1. Move migration target from merged-packages/ to packages/.
2. Run yarn install in the root directory.
3. Check that all tests are passing in migration target by running `yarn workspace @metamask/json-rpc-middleware-stream test`.

Lints and Constraints
1. Fixed `yarn lint` errors
2. Fixed `yarn constraints` errors
3. Applied `yarn workspace @metamask/json-rpc-middleware-stream changelog:validate`

## References

- Closes #1920 

### `@metamask/json-rpc-middleware-stream`

- **BREAKING:** Rename package from `json-rpc-middleware-stream` to `@metamask/json-rpc-middleware-stream`

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

**Comment by mcmire** [comment]
Per standup, we should take a look at the [migration checklist](https://github.com/MetaMask/core/issues/1551#issuecomment-1745665740) for this package again as it may have changed since this PR was created. If there are any PRs we need to merge before this one can go in then we should take care of that first.

**Review by MajorLift — CHANGES_REQUESTED** [review]
I've finished testing the tag porting script, and I think it will need to be run before the package is moved out of `merged-packages/`. So https://github.com/MetaMask/core/pull/1802 will need to be merged before this PR can be merged. Sorry for the confusion!

File: packages/json-rpc-middleware-stream/package.json
Comment (mcmire) [inline packages/json-rpc-middleware-stream/package.json:3]: Hmm, I'm not sure we should bump the version yet. Typically we only bump packages in a release PR. Given that other teams can create release PRs whenever they want to, if we do this now, this new version may be published without our knowledge. What do you think about waiting to do this? I believe the change to `changelog:validate` can stay the same though (unless I'm missing something).

File: packages/json-rpc-middleware-stream/tsconfig.json
Comment (mcmire) [inline packages/json-rpc-middleware-stream/tsconfig.json:12]: Consulting the [original repo](https://github.com/MetaMask/json-rpc-middleware-stream/blob/main/tsconfig.json), it doesn't look like `exactOptionalPropertyTypes`, `noErrorTruncation, or `noUncheckedIndexedAccess` are present, and the `target` seems to be set to `es2017` (instead of `es6`, which is the default for this repo). `skipLibCheck` _is_ set, but we could drop that if it's not causing any problems.

Additionally, consulting other packages in this repo, e.g. [`accounts-controller`](https://github.com/MetaMask/core/blob/5ff1505aa4833f7b6c9055cc734c18be2401b661/packages/accounts-controller/tsconfig.json), `rootDir` doesn't need to be set, and `include` doesn't need to include `tests`.

What do you think about:

``` suggestion
  "compilerOptions": {
    "baseUrl": "./",
    "forceConsistentCasingInFileNames": true,
    "target": "es2017"
  },
  "include": ["../../types", "./src"]

File: packages/json-rpc-middleware-stream/package.json
Comment (mcmire) [inline packages/json-rpc-middleware-stream/package.json:40]: Nit: It looks like this is ^2.3.9 in the [original repo](https://github.com/MetaMask/json-rpc-middleware-stream/blob/main/package.json):

``` suggestion
    "@types/readable-stream": "^2.3.9",

File: packages/json-rpc-middleware-stream/src/createStreamMiddleware.ts
Comment (mcmire) [inline packages/json-rpc-middleware-stream/src/createStreamMiddleware.ts:153]: Nit: To make this diff smaller, can you also say this as this?

``` suggestion
      idMap[req.id]?.retryCount = retryCount + 1;

**Review by mcmire — COMMENTED** [review]
I had a few comments, but otherwise this looks good to me.

File: packages/json-rpc-middleware-stream/src/createStreamMiddleware.ts
Comment (MajorLift) [reply packages/json-rpc-middleware-stream/src/createStreamMiddleware.ts:153]: I think optional chaining can't be used on the LHS of an assignment.

File: packages/json-rpc-middleware-stream/tsconfig.json
Comment (MajorLift) [reply packages/json-rpc-middleware-stream/tsconfig.json:12]: I agree, we should only be preserving the compiler options of the original, not introducing new ones.

File: packages/json-rpc-middleware-stream/package.json
Comment (MajorLift) [reply packages/json-rpc-middleware-stream/package.json:3]: This one's a little tricky. Usually the first post-migration release should happen in core as the last step in the migration checklist, but in this case the complication is that `json-rpc-middleware-stream` has never been renamed into the `@metamask/` namespace. 

I think ideally the rename and release should happen in the original repo before git history migration (I'll add an entry about this in the checklist). However, in this case, that essentially means restarting the migration. I don't think doing the rename+release at this stage is safe or desirable, though. 

We could do it as part of the first post-migration release in core. I'll have to think about whether that could introduce any issues.

File: packages/json-rpc-middleware-stream/package.json
Comment (MajorLift) [reply packages/json-rpc-middleware-stream/package.json:40]: Yes this is enforced by yarn constraints. Setting this to "^2.3.9" triggers an error because the least (and only) version used in the monorepo is "^2.3.0". 

If using the latest version for a certain dependency is important, we need to open a PR in core to bump it throughout the monorepo first. Otherwise, this needs to be decremented for now.

File: packages/json-rpc-middleware-stream/package.json
Comment (MajorLift) [reply packages/json-rpc-middleware-stream/package.json:3]: One issue of doing the rename post-migration is that yarn constraints requires all subpackages to have `@metamask/`-namespaced "name" fields in their package.json, so we can't merge this PR without this change.

File: packages/json-rpc-middleware-stream/tsconfig.json
Comment (mcmire) [reply packages/json-rpc-middleware-stream/tsconfig.json:12]: Ah. Good eye! That is because they reference type augmentations — type overrides for packages — that are defined _outside_ of the package. If we didn't define `"rootDir": "../.."` in those cases then TypeScript would give us an error. In this case, it doesn't look like we are accessing those types, so we don't need to specify a `rootDir`. Does that make sense? (Perhaps we ought to be more consistent about this.)

File: packages/json-rpc-middleware-stream/package.json
Comment (mcmire) [reply packages/json-rpc-middleware-stream/package.json:40]: Oh, I swear I searched the repo for this package. I see now it's used by `accounts-controller`. Okay, looks good then!

File: packages/json-rpc-middleware-stream/src/createStreamMiddleware.ts
Comment (mcmire) [reply packages/json-rpc-middleware-stream/src/createStreamMiddleware.ts:153]: Thank you 🤦🏻

File: packages/json-rpc-middleware-stream/package.json
Comment (mcmire) [reply packages/json-rpc-middleware-stream/package.json]: Ah sorry, I forgot that we were doing a rename here. Well, we'll just have to keep an eye on releases then, no problem.

**Review by MajorLift — APPROVED** [review]
Tags have been successfully ported to core! Verify with this command:
```bash
git ls-remote --tags origin | grep 'json-rpc-middleware-stream'
```

**Comment by MajorLift** [comment]
This package was renamed from `json-rpc-middleware-stream` to `@metamask/json-rpc-middleware-stream` in this PR. This change will be reflected in the next core release that includes this package. 

An "unreleased" CHANGELOG entry will be added as part of the post-migration cleanup that will follow: 
- Change package name to `@metamask/json-rpc-middleware-stream` ([#162](https://github.com/MetaMask/core/pull/1762))

File: packages/permission-controller/package.json
Comment (FrederikBolding) [inline packages/permission-controller/package.json:38]: Why is this downgrading these packages? 🤔

Comment (MajorLift) [reply packages/permission-controller/package.json]: Thanks for catching this! https://github.com/MetaMask/core/pull/1895#discussion_r1381269939

**Comment by mcmire** [comment]
> This package was renamed from json-rpc-middleware-stream to @metamask/json-rpc-middleware-stream in this PR.
>
> [...]
>
> An "unreleased" CHANGELOG entry will be added as part of the post-migration cleanup that will follow

I've added a changelog update to the PR description just in case anyone is looking at this PR.

We need to be super cautious about the next core releases to ensure that this package does not get published accidentally before we bump the version number.

**Review by mcmire — APPROVED** [review]
This looks good to me!

**Comment by MajorLift** [comment]
- Unable to repro [lint error](https://github.com/MetaMask/core/actions/runs/6777049684/job/18419905745?pr=1762) on local (see https://github.com/MetaMask/core/issues/1989).
- Suppressed `@typescript-eslint/no-implied-eval` warning: https://github.com/MetaMask/core/pull/1762/commits/7f5e3b5688c06be76b6f8dc44d75e0ae6b71554f

**Review by mcmire — APPROVED** [review]
LGTM.

**Pull request merged**
